### PR TITLE
refact related to `new Expressions` and the array dim

### PR DESCRIPTION
**Pull request by ghost**
- when possible use the known dim as ctor arg and rewrite the following `push` as index assign.
- use a `foreach` loop if dim was already set in the exprs ctor but a `for` loop still used.
- use a cleaner `foreach(ref elem; *expr)` when possible.

### CODE REVIEW

File: src/dmd/cond.d [hunk 1]
```diff
@@ -149,13 +149,13 @@ extern (C++) final class StaticForeach : RootObject
         el = el.ctfeInterpret();
         if (el.op == TOK.int64)
         {
-            dinteger_t length = el.toInteger();
-            auto es = new Expressions();
-            foreach (i; 0 .. length)
+            const length = cast(size_t) el.toInteger();
```
Comment (Geod24) [inline src/dmd/cond.d]: Doesn't work with cross compiler

Comment (Basile-z) [reply src/dmd/cond.d]: what is the name of the target size_t type ?

Comment (thewilsonator) [reply src/dmd/cond.d]: `d_targ_size_t` or something like that.

Comment (Basile-z) [reply src/dmd/cond.d]: I can assert on the length before casting. Look at the context, such an amount of element would create an OOM anyway.

Comment (Basile-z) [reply src/dmd/cond.d]: Finally removed this change instead of speculating on what would work or not.

**Review by thewilsonator — APPROVED** [review]

**Comment by jacob-carlborg** [comment]
Hmm, I would have added an overload to the constructor to take a variadic number of `T` and hide the implementation details inside `Array`.

**Comment by Basile-z** [comment]
Yeah, but the problem is that when T is an integer type. Is the parameter an element of a dimension ?

**Comment by jacob-carlborg** [comment]
> Yeah, but the problem is that when T is an integer type. Is the parameter an element of a dimension?

I didn't think of that. Is the dimension overload useful if the other one exists?

**Comment by Basile-z** [comment]
never mind, it's just fine.

**Pull request merged**